### PR TITLE
coll: Add GPU fallback for other op-based collectives

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Follow on to #4772. Takes care of other op-based collectives.

## Expected Impact

Enable GPU support for these collectives.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by tarudoodi** [comment]
@raffenet will you update the relevant tests in mpich testsuite too?

**Comment by raffenet** [comment]
> @raffenet will you update the relevant tests in mpich testsuite too?

Yes, I suppose that would be a good thing. Will update the PR.

**Comment by raffenet** [comment]
Rather than modify a bunch of existing tests, I created an op-based test for all blocking and nonblocking collectives. It is adapted from `nonblocking2.c`.
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by tarudoodi** [comment]
> Rather than modify a bunch of existing tests, I created an op-based test for all blocking and nonblocking collectives. It is adapted from `nonblocking2.c`.

I think that should be OK, since most of the non-blocking tests are simpler and run with `MPI_COMM_WORLD`.

If we look at it from the point of view of testing different algorithms `test_coll_algos.sh`:
1. There is an algorithm for `Scan` and `Iscan` that gets tested there. Do you think it would be ok to add `op_coll` to that testing?
2. Now that `MPI_Iallreduce`, `MPI_Ireduce` support fallback path for gpu buffers we can add the options for different `memtype`s for gentran algorithm testing.  They use regular blocking reduce and allreduce tests and force the use of non-blocking algorithms.

**Comment by raffenet** [comment]
> > Rather than modify a bunch of existing tests, I created an op-based test for all blocking and nonblocking collectives. It is adapted from `nonblocking2.c`.
> 
> I think that should be OK, since most of the non-blocking tests are simpler and run with `MPI_COMM_WORLD`.
> 
> If we look at it from the point of view of testing different algorithms `test_coll_algos.sh`:
> 
> 1. There is an algorithm for `Scan` and `Iscan` that gets tested there. Do you think it would be ok to add `op_coll` to that testing?

Sure, no problem.

> 2. Now that `MPI_Iallreduce`, `MPI_Ireduce` support fallback path for gpu buffers we can add the options for different `memtype`s for gentran algorithm testing.  They use regular blocking reduce and allreduce tests and force the use of non-blocking algorithms.

Do you mean generate `testlist.gpu` entries for nonblocking reduce and allreduce tests in the `test_coll_algos.sh` script?

**Comment by tarudoodi** [comment]
> Do you mean generate `testlist.gpu` entries for nonblocking reduce and allreduce tests in the `test_coll_algos.sh` script?

Yes.

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Comment by raffenet** [comment]
@tarudoodi pushed 2 new commits. Let me know if this is along the lines you were thinking.

**Comment by tarudoodi** [comment]
> @tarudoodi pushed 2 new commits. Let me know if this is along the lines you were thinking.

Yes, that's what I had in mind. Should we add `memtype`s to `op_coll` for `scan` test too? I feel it may not be that important, the ones from `generate_gpu_tests.sh` should be sufficient. I have not worked much on `MPI_Scan` so I may be wrong.

For `MPI_Iallreduce`, `allred2` can also support the different `memtype`s. I think key difference from `allred` would be that it uses an `in_place` sendbuf. Do you think we should add that too?

**Comment by raffenet** [comment]
> > @tarudoodi pushed 2 new commits. Let me know if this is along the lines you were thinking.
> 
> Yes, that's what I had in mind. Should we add `memtype`s to `op_coll` for `scan` test too? I feel it may not be that important, the ones from `generate_gpu_tests.sh` should be sufficient. I have not worked much on `MPI_Scan` so I may be wrong.

I think at the moment our support is so basic that it probably does not matter. But there's not much effort to add it, so I might as well.

> 
> For `MPI_Iallreduce`, `allred2` can also support the different `memtype`s. I think key difference from `allred` would be that it uses an `in_place` sendbuf. Do you think we should add that too?

Yes, it will be good to catch `MPI_IN_PLACE`. Will add.

**Comment by tarudoodi** [comment]
This looks good and complete to me now :)

File: src/mpi/coll/scan/scan.c [hunk 1]
```diff
@@ -153,6 +162,14 @@ int MPIR_Scan(const void *sendbuf, void *recvbuf, int count,
         mpi_errno = MPIR_Scan_impl(sendbuf, recvbuf, count, datatype, op, comm_ptr, errflag);
     }
 
+    /* Copy out data from host recv buffer to GPU buffer after reduce is complete */
```
Comment (tarudoodi) [inline src/mpi/coll/scan/scan.c:165]: One minor  typo here and other similar places, it still says `reduce` .

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

File: src/mpi/coll/scan/scan.c (hunk 1 above)
Comment (raffenet) [reply src/mpi/coll/scan/scan.c]: Fixed.

**Review by tarudoodi — APPROVED** [review]

**Pull request merged**